exporter/trace/stackdriver/trace.go

```diff
@@ -0,0 +1,182 @@
+		e.uploadFn(bundle.([]*tracepb.Span))
+	})
+	if o.BundleDelayThreshold > 0 {
+		b.DelayThreshold = o.BundleDelayThreshold
+	} else {
+		b.DelayThreshold = 2 * time.Second
+	}
+	if o.BundleCountThreshold > 0 {
+		b.BundleCountThreshold = o.BundleCountThreshold
+	} else {
+		b.BundleCountThreshold = 50
+	}
+	// The measured "bytes" are not really bytes, see exportReceiver.
+	b.BundleByteThreshold = b.BundleCountThreshold * 200
+	b.BundleByteLimit = b.BundleCountThreshold * 1000
+	if o.TraceSpansBufferMaxBytes > 0 {
+		b.BundleByteLimit = o.TraceSpansBufferMaxBytes
+	} else {
+		b.BundleByteLimit = defaultBufferedByteLimit
+	}
+
+	e.bundler = b
+	e.uploadFn = e.uploadSpans
+	return e
+}
+
+// ExportSpan exports a SpanData to Stackdriver Trace.
+func (e *traceExporter) ExportSpan(s *trace.SpanData) {
+	protoSpan := protoFromSpanData(s, e.projectID)
+	protoSize := proto.Size(protoSpan)
+	err := e.bundler.Add(protoSpan, protoSize)
+	switch err {
+	case nil:
+		return
+	case bundler.ErrOversizedItem:
+	case bundler.ErrOverflow:
+		e.overflowLogger.log()
+	default:
+		e.o.handleError(err)
+	}
+}
+
+// Flush waits for exported trace spans to be uploaded.
+//
+// This is useful if your program is ending and you do not want to lose recent
+// spans.
+func (e *traceExporter) Flush() {
+	e.bundler.Flush()
+}
+
+// uploadSpans uploads a set of spans to Stackdriver.
+func (e *traceExporter) uploadSpans(spans []*tracepb.Span) {
+	req := tracepb.BatchWriteSpansRequest{
+		Name:  "projects/" + e.projectID,
+		Spans: spans,
+	}
+	// Create a never-sampled span to prevent traces associated with exporter.
+	ctx, cancel := newContextWithTimeout(e.o.Context, e.o.Timeout)
+	defer cancel()
+
+	// TODO(ymotongpoo): add this part after OTel support NeverSampler
+	// for tracer.Start() initialization.
+	//
+	// tracer := apitrace.Register()
+	// ctx, span := tracer.Start(
+	// 	ctx,
+	// 	"go.opentelemetry.io/exporter/stackdriver.uploadSpans",
+	// )
+	// defer span.End()
+	// span.SetAttribute(key.New("num_spans").Int64(int64(len(spans))))
+
+	err := e.client.BatchWriteSpans(ctx, &req)
+	if err != nil {
+		// TODO(ymotongpoo): handle detailed error categories
+		// span.SetStatus(codes.Unknown)
+		e.o.handleError(err)
+	}
+}
+
+// overflowLogger ensures that at most one overflow error log message is
+// written every 5 seconds.
+type overflowLogger struct {
+	mu    sync.Mutex
+	pause bool
+	accum int
+}
+
+func (o *overflowLogger) delay() {
+	o.pause = true
+	time.AfterFunc(5*time.Second, func() {
+		o.mu.Lock()
+		defer o.mu.Unlock()
+		switch {
+		case o.accum == 0:
+			o.pause = false
+		case o.accum == 1:
+			log.Println("OpenTelemetry Stackdriver exporter: failed to upload span: buffer full")
+			o.accum = 0
+			o.delay()
+		default:
+			log.Printf("OpenTelemetry Stackdriver exporter: failed to upload %d spans: buffer full", o.accum)
+			o.accum = 0
+			o.delay()
+		}
+	})
+}
+
+func (o *overflowLogger) log() {
+	o.mu.Lock()
```

Comment:
It looks dangerous to `log.Println` while holding a lock, I'd say.

Reply:
Eventually, overflowLogger was totally removed as a result of elimination of bundler.